### ISSUE TITLE
Help with finding python library issue in manylinux CI build?

### ISSUE THREAD

**Issue by makaimann** [joerick/cibuildwheel#477]
Thank you for all your work on this tool! It's a very useful and cool idea.

I've been working on incorporating `cibuildwheel` into our GitHub Actions CI flow for the [CVC4](https://github.com/CVC4/CVC4) python bindings (e.g. a binary extension wheel) in my [fork](https://github.com/makaimann/CVC4/tree/pypi). I'm using [scikit-build](https://github.com/scikit-build/scikit-build) for the `setup.py` infrastructure, and `Cython` for the bindings. Unfortunately, I've run into an issue for the manylinux build where it cannot find the python library.

Although I recognize that this is somewhat independent of `cibuildwheel`, I thought I'd ask about it here because I can't reproduce this in the docker image manually. It works fine if I pull the `quay.io/pypa/manylinux2010_x86_64` docker image myself, set it up with `rh-python36` and then run my `before_all` script followed by `pip wheel . -w wheelhouse`. However it fails in `cibuildwheel` with both `manylinux2010` and `manylinux2014` (logs [here](https://github.com/makaimann/CVC4/runs/1500806063?check_suite_focus=true) and [here](https://github.com/makaimann/CVC4/runs/1513064399?check_suite_focus=true)).

The error I get is: 
```
[100%] Linking CXX shared module pycvc4.so
/opt/rh/devtoolset-8/root/usr/libexec/gcc/x86_64-redhat-linux/8/ld.gold: error: cannot find -lpython3.6m
/opt/rh/devtoolset-8/root/usr/libexec/gcc/x86_64-redhat-linux/8/ld.gold: error: cannot find -lpython3.6m
```

Based on printouts it was configured with:
```
cmake /tmp/pip-req-build-gagwvkwh -G 'Unix Makefiles' -DCMAKE_INSTALL_PREFIX:PATH=/tmp/pip-req-build-gagwvkwh/_skbuild/linux-x86_64-3.6/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/opt/_internal/cpython-3.6.12/bin/python3.6 -DPYTHON_VERSION_STRING:STRING=3.6.12 -DPYTHON_INCLUDE_DIR:PATH=/opt/_internal/cpython-3.6.12/include/python3.6m -DPYTHON_LIBRARY:FILEPATH=libpython3.6m.a -DSKBUILD:BOOL=TRUE -DCMAKE_MODULE_PATH:PATH=/tmp/pip-build-env-19nmy1bj/overlay/lib/python3.6/site-packages/skbuild/resources/cmake -DBUILD_BINDINGS_PYTHON=ON -DBUILD_LIB_ONLY=ON -DCMAKE_BUILD_TYPE:STRING=Release
```

Do you by any chance have advice on how to best debug this? In particular, I'm curious about the difference between `cibuildwheel` and the environment in the docker image I'm pulling and running manually.

**Comment by henryiii** [comment]
First, very quick thought: why are you linking to Python? You should not be statically linking to Python on Linux or macOS. When you import an extension module, the symbols from Python are already loaded, and if you mix patch versions, it could crash if you try to import the symbols again. Basically this line is quite illegal on non-Windows systems:

https://github.com/makaimann/CVC4/blob/c2f97c27014869a68a8c3591e69ca4324f70f976/src/api/pycvc4/CMakeLists.txt#L49

**Comment by henryiii** [comment]
PS: These static files are deleted from the manylinux images to save space and to avoid building buggy wheels - so not sure how you managed to build manually inside the image.

**Comment by Czaki** [comment]
Did your browser also freeze when trying to unfold logs?

Maybe in 

> set it up with `rh-python36` and then run my

you install missing files?

**Comment by makaimann** [comment]
Thank you both for the quick response -- it's much appreciated. 

@henryiii, that's a good point. To be honest, I have no idea why that's there and it doesn't seem necessary so I just removed that. We'll see if it changes anything. For what it's worth, I'm not sure that it's statically linking (at least not always). On my Mac, `PYTHON_LIBRARIES` is pointing to a dynamic library. Regardless this is very good to know and I won't link against it anymore.

@Czaki my browser handles them, but the freezing is probably due to the `-Wshadow` warnings in Cython-generated code -- sorry about that. If you're interested, raw logs are attached. The interesting ones are the ubuntu builds in the "Build wheels" action. That would make sense if it's related to `rh-python36`. How is the python version handled by `cibuildwheel` and could I supplement it with missing files somehow?

I'll respond when the latest GitHub Actions complete in case the linking change already solved this.

[logs_manylinux2010.tar.gz](https://github.com/joerick/cibuildwheel/files/5655887/logs_manylinux2010.tar.gz)
[logs_manylinux2014.tar.gz](https://github.com/joerick/cibuildwheel/files/5655888/logs_manylinux2014.tar.gz)

**Comment by henryiii** [comment]
Ahh, didn't read that. So yes, you are not using manylinux Python so that's why it worked. That's no different than just using a normal RHEL based image like CentOS. Look at the manylinux example to see how to use the manylinux Pythons, basically it is by using the full paths in /opt

> Did your browser also freeze when trying to unfold logs

No. Though my MacBook also has 128GB of memory. Text shouldn't be a resource hog, but it is a browser...

**Comment by Czaki** [comment]
> > Did your browser also freeze when trying to unfold logs
> 
> No. Though my MacBook also has 128GB of memory. Text shouldn't be a resource hog, but it is a browser...

I still have over 10 GB of RAM free, but tab hang. looks like js problem when unfold... 


> How is the python version handled by `cibuildwheel` and could I supplement it with missing files somehow?

You may use yum for install tools from repository. You could use custom docker image. But remember that it need pass auditwheel, which may block part of system libraries. See for example https://www.python.org/dev/peps/pep-0571/#the-manylinux2010-policy

**Comment by henryiii** [comment]
> PYTHON_LIBRARIES is pointing to a dynamic library

Still do not link to it. You should use `-undefined,dynamic_lookup` on macOS, see https://gitlab.kitware.com/cmake/cmake/-/blob/e94f00dfe3c5a4f357a34a37d397c1fbbf1f52d5/Modules/FindPython/Support.cmake#L3270 for example. Scikit-build's integrated CMake module has a [fancy version](https://github.com/scikit-build/scikit-build/blob/c0741f9a1de9dd97e0a1e6f0b3b02fcd76221418/skbuild/resources/cmake/FindPythonExtensions.cmake#L427) of this that only makes the symbols in the Python library undefined dynamic lookup, if you wanted to use `python_extension_module` from `find_package(PythonExtensions)`, which handles all this stuff for you.

**Comment by henryiii** [comment]
By the way, [PEP 600](https://www.python.org/dev/peps/pep-0600/) is a great read on packaging, even though its focus is something new that was just released in pip 20.3.

**Comment by YannickJadoul** [comment]
> So yes, you are not using manylinux Python so that's why it worked.

2 extra - superfluous - cents: all manylinux images already have Python versions installed under `/opt/python/...`, so no need to install that, normally.

**Comment by makaimann** [comment]
Thanks everyone for the advice! I'll check out PEP 600, and it was good to learn more about manylinux / to avoid linking against the python libraries. I still haven't gotten it fully working, but I'm through the build now. `auditwheel` is failing but I suspect it has to do with `scikit-build` automatically cleaning up build files -- need to look into that still. Regardless, that is beyond the scope of this thread.

For other people who might happen upon this thread: the issue was indeed that there's no need to link against the python libraries and as far as I can tell, the `manylinux` python versions don't have the libraries bundled with them. Removing the explicit linking helped, and then it also turned out that we were using an old version of the `UseCython` and related cmake files (probably from an old version of `scikit-build`). If I recall correctly, there was at some point an issue with `target_link_libraries_with_dynamic_lookup` that has now been resolved. I ran into that before without fully appreciating its purpose. So, updating those cmake files was the final necessary step.

Thanks again!
